File: packages/honker-ex/lib/honker/extension.ex
defmodule Honker.Extension do
  @moduledoc """
  Locates the Honker SQLite loadable extension.

  Hex packages cannot ship a compiled shared library the way the
  Python, Ruby, .NET, and JVM packages do, so unlike those bindings
  there is nothing bundled to fall back on. Set `HONKER_EXTENSION_PATH`,
  or download the extension from a Honker release:

      https://github.com/russellromney/honker/releases

  Use it to fill in `:extension_path` for `Honker.open/2`, or to load
  Honker onto an Exqlite connection you already own — which is what you
  want with Ecto, since enqueueing outside your transaction loses
  atomicity.

      {:ok, path} = Honker.Extension.path()
      Honker.open("app.db", extension_path: path)
  """

  @typedoc "Absolute path to the extension shared library."
  @type path :: String.t()

  @doc """
  The SQLite entry point exported by the extension.

  SQLite normally derives this from the file name — strip a leading
  `lib`, take characters up to the first `.`, keep the alphabetic ones,
  so `libhonker_ext.so` gives `honkerext`. It only matters when loading
  the extension under some other name.
  """
  @spec entrypoint() :: String.t()
  def entrypoint, do: "sqlite3_honkerext_init"

  @doc "The extension's file name on this platform."
  @spec filename() :: String.t()
  def filename do
    case :os.type() do
      {:win32, _} -> "honker_ext.dll"
      {:unix, :darwin} -> "libhonker_ext.dylib"
      _ -> "libhonker_ext.so"
    end
  end

  @doc """
  Locate the extension.

  Resolution order is `HONKER_EXTENSION_PATH`, then the current working
  directory, then `priv/` of the `:honker` application. Returns
  `{:error, message}` naming every path tried when none exists; nothing
  is guessed.
  """
  @spec path() :: {:ok, path()} | {:error, String.t()}
  def path do
    case System.get_env("HONKER_EXTENSION_PATH") do
      nil ->
        search()

      "" ->
        search()

      env ->
        if file?(env),
          do: {:ok, env},
          else: {:error, "HONKER_EXTENSION_PATH does not exist: #{env}"}
    end
  end

  @doc """
  Same as `path/0` but raises instead of returning an error tuple.
  """
  @spec path!() :: path()
  def path! do
    case path() do
      {:ok, p} -> p
      {:error, message} -> raise ArgumentError, message
    end
  end

  defp search do
    candidates = Enum.map(search_dirs(), &Path.join(&1, filename()))

    case Enum.find(candidates, &file?/1) do
      nil ->
        {:error,
         "Honker SQLite extension not found. The Elixir binding ships no binary — " <>
           "set HONKER_EXTENSION_PATH, or download honker-ext-<target>.tar.gz from " <>
           "https://github.com/russellromney/honker/releases and extract " <>
           "#{filename()} from it, or build it with " <>
           "`cargo build --release -p honker-extension`. Keep the file name as-is: " <>
           "SQLite derives the extension entry point from it. Searched: " <>
           Enum.join(candidates, ", ")}

      found ->
        {:ok, found}
    end
  end

  defp search_dirs do
    priv =
      case :code.priv_dir(:honker) do
        {:error, _} -> []
        dir -> [List.to_string(dir)]
      end

    [File.cwd!() | priv]
  end

  defp file?(path), do: File.regular?(path)
end

File: packages/honker-go/extension.go
package honker

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

// ExtensionEntrypoint is the SQLite entry point exported by the Honker
// extension.
//
// When no entry point is given, SQLite derives one from the file name.
// That works for the canonical libhonker_ext.{so,dylib} /
// honker_ext.dll. Pass this explicitly if the library is named anything
// else — the derivation is version-dependent.
const ExtensionEntrypoint = "sqlite3_honkerext_init"

// ExtensionFilename is the extension's file name on this platform.
func ExtensionFilename() string {
	switch runtime.GOOS {
	case "windows":
		return "honker_ext.dll"
	case "darwin":
		return "libhonker_ext.dylib"
	default:
		return "libhonker_ext.so"
	}
}

// ExtensionPath locates the Honker SQLite extension.
//
// Go modules cannot ship a compiled binary, so unlike the Python, Ruby,
// .NET, and JVM packages there is nothing bundled to fall back on. Set
// HONKER_EXTENSION_PATH, or download the extension from a Honker
// release:
//
//	https://github.com/russellromney/honker/releases
//
// Resolution order is HONKER_EXTENSION_PATH, then the extension next to
// the running binary, then the working directory. A miss returns an
// error naming every path tried; nothing is guessed.
//
// Use it to fill in the extensionPath argument to Open, or to load
// Honker onto a *sql.DB you already own.
func ExtensionPath() (string, error) {
	if env := os.Getenv("HONKER_EXTENSION_PATH"); env != "" {
		if isFile(env) {
			return env, nil
		}
		return "", fmt.Errorf("honker: HONKER_EXTENSION_PATH does not exist: %s", env)
	}

	name := ExtensionFilename()
	var searched []string

	if exe, err := os.Executable(); err == nil {
		searched = append(searched, filepath.Join(filepath.Dir(exe), name))
	}
	if cwd, err := os.Getwd(); err == nil {
		searched = append(searched, filepath.Join(cwd, name))
	}

	for _, candidate := range searched {
		if isFile(candidate) {
			return candidate, nil
		}
	}

	return "", fmt.Errorf(
		"honker: SQLite extension not found. The Go binding ships no binary — "+
			"set HONKER_EXTENSION_PATH, or download honker-ext-<target>.tar.gz "+
			"from https://github.com/russellromney/honker/releases and extract "+
			"%s from it, or build it with "+
			"`cargo build --release -p honker-extension`. Keep the file name as-is: "+
			"SQLite derives the extension entry point from it. Searched: %s",
		name,
		strings.Join(searched, ", "),
	)
}

func isFile(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.Mode().IsRegular()
}
